### ISSUE TITLE
Defer code execution and unblock the main thread while maintaining payload accuracy @snowplow/browser-tracker | core

### ISSUE THREAD

**Issue by dg-j7i** [snowplow/snowplow-javascript-tracker#1289]
We are currently optimizing our single page application towards faster visual feedback on user interactions.

Within our application, the browser has to do (a lot of) work when a user interacts with the page and e.g. performs a click. This includes capturing user behavior by running tracking events and blocking the main thread while doing so.

Currently, when we schedule an event to be tracked with `trackSelfDescribingEvent()` from the `@snowplow/browser-tracker` library, we provide information about the event to be tracked:
We pass event atomics and custom context data when calling the tracking event. But we are not able to build and provide the full payload when making this call. So later on the library will enhance the tracking payload with additional (contextual) information, for example the `page_url` the event was sent from (as of our current understanding).

The current implementation of this tracking function does not allow us to schedule a tracking event with all the necessary information collected upfront and thus become independent of any delays in its execution.

![image](https://github.com/snowplow/snowplow-javascript-tracker/assets/40632502/559d923e-4a53-4e85-a910-1c4f083f8c2e)

### The solution we'd like
We would like to be able to defer event tracking executions while maintaining event payload accuracy, by providing the entire payload upfront when we schedule the call.

### Alternatives we've considered
We have considered identifying parts of the payload that change between the scheduling and execution of events and possibly providing them additionally as custom properties or context. This would unnecessarily bloat the payload with duplicate and sometimes incorrect information, and also cause confusion when processing the data.

### Additional context
We are open to contributing and/or supporting further by providing more specifics and/or resources.
We were also wondering if this is already possible and just not documented or has been overseen.

https://web.dev/articles/optimize-long-tasks

**Comment by rvetere** [comment]
After analysing more performance profiles, it becomes clear what is "expensive" about the current tracking solution:

Every `beforeTrack` call will enrich the payload builder with so called browser properties. A few of them like `viewport`, `documentSize` and `screenResolution` are expensive calls.

Why are they expensive? Because they involve properties like `innerWidth`, `innerHeight`, `clientWidth`, `clientHeight` and `scrollHeight`. If you try to use one of these, the browser has to actually calculate the size and thus will execute `recalculateStyles` to determine the current pixel value - this has a certain cost and will be much more time in complex web applications.

It is clearly visible in a performance profile too:
![Screenshot 2024-03-06 at 14 22 10](https://github.com/snowplow/snowplow-javascript-tracker/assets/1505976/b51559bc-3524-4edf-b2d7-586807dc942c)

-> The purple "Recalculate Style" block should not be there underneath `beforeTrack`

## TLDR;

I'm currently working on a change that would improve this behavior by using a `ResizeObserver` and not grabbing these properties synchronously on every `track()` call

**Comment by rvetere** [comment]
Here would be my PR to fix this behavior :)
https://github.com/snowplow/snowplow-javascript-tracker/pull/1294

**Comment by dg-j7i** [comment]
@igneel64 Do you may have any other alternatives in mind which could help us mitigate on these performance issues we're facing currently? What do you think about the mentioned issue?

The changes suggested by @rvetere would drastically minimize the performance overhead; although they would not completely decouple e.g. click tracking from the handling of clicks in terms of ui updates.

**Comment by matus-tomlein** [comment]
First of all, thank you @dg-j7i for raising this issue and @rvetere for contributing the PR! Sorry for the delay, we will give this a deep review next week, discuss in the team and come back to you!

**Comment by matus-tomlein** [comment]
@dg-j7i @rvetere an update from our side – we have discussed the proposed approach (the one in the PR) in the team and we are in line with going in this direction! We're still testing this though, will add a review in the PR next week.

**Comment by dg-j7i** [comment]
I see the [javascript-tracker package](https://www.npmjs.com/package/@snowplow/javascript-tracker) is published and up to date. 🎉 
Will the update fort the [browser-tracker package](https://www.npmjs.com/package/@snowplow/browser-tracker) also be available soon?

We look forward to seeing the performance improvements reflected in our web vitality metrics.
Thanks again

**Comment by matus-tomlein** [comment]
Hi @dg-j7i, the browser-tracker should already be updated on NPM, look for version 3.23.0.

The thanks goes to you! Thanks for raising this and contributing the solution! Happy to hear any feedback you might have about this.

**Comment by matus-tomlein** [comment]
Closing as this was released in version 3.23.0.
